### PR TITLE
CDAP-5605 fix incorrect aggregator metrics

### PR DESCRIPTION
**Pull request by albertshau**
Fixed a bug where the metrics for records in and out of an
aggregator were incorrect because they were counting intermediate
group metrics.

Includes some refactoring so that metrics are all handled in
TrackedTransform instead of partly in TrackedTransform and partly
in DefaultEmitter. Also making the tracking classes a little more
generic so that the metric emitted can be set. This allows us to
introduce a new 'aggregator.groups' metric to count the number of
groups in an aggregator.

### CODE REVIEW

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-DUT4012

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkTransformExecutorFactory.java [hunk 1]
```diff
@@ -51,18 +52,19 @@ public void transform(Object input, Emitter emitter) throws Exception {
   private final PluginContext pluginContext;
   private final long logicalStartTime;
   private final Map<String, String> runtimeArgs;
-  private final boolean isPreGroup;
+  // whether this is before the group step of an aggregator, or before a spark compute stage
+  private final boolean isBeforeBreak;
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkTransformExecutorFactory.java]: why the name `break` though ? hard to understand.

Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkTransformExecutorFactory.java]: yeah i wasn't very happy with this name either.  I may just see if I can refactor the spark compute code path

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/TransformExecutorFactory.java [hunk 2]
```diff
@@ -89,4 +90,14 @@ protected Transformation getTransformation(String pluginType, String stageName)
...
   }
 
+  protected static <IN, OUT> TrackedTransform<IN, OUT> getTrackedGroupStep(Transformation<IN, OUT> transform,
+                                                                              StageMetrics stageMetrics) {
+    return new TrackedTransform<>(transform, stageMetrics, TrackedTransform.RECORDS_IN, null);
+  }
+
+  protected static <IN, OUT> TrackedTransform<IN, OUT> getTrackedAggregateStep(Transformation<IN, OUT> transform,
+                                                                               StageMetrics stageMetrics) {
+    return new TrackedTransform<>(transform, stageMetrics, "aggregator.groups", TrackedTransform.RECORDS_OUT);
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/TransformExecutorFactory.java]: add a comment about what aggregator.groups is, why we need them.

Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/TransformExecutorFactory.java]: should we make this generic and not have two (trackedgroup, trackedaggregate) separate methods ?

**Comment by gsps1** [comment]
lgtm

**Comment by albertshau** [comment]
changes were just to add a comment and rename a variable.  Last build was green so merging.

**Pull request merged**